// react-mixins/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.configMixin = exports.customConfigMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var customConfigMixin = exports.customConfigMixin = function customConfigMixin() {
    var _ref = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    var _ref$paramName = _ref.paramName;
    var paramName = _ref$paramName === undefined ? 'config' : _ref$paramName;
    var _ref$prioritizeNamed = _ref.prioritizeNamed;
    var prioritizeNamed = _ref$prioritizeNamed === undefined ? false : _ref$prioritizeNamed;

    return {
        propTypes: _defineProperty({}, paramName, _react2.default.PropTypes.object),
        spreadConfig: function spreadConfig() {
            var _props = this.props;
            var config = _props.config;

            var other = _objectWithoutProperties(_props, ['config']);

            this.props = prioritizeNamed ? Object.assign(other, config) : Object.assign(config, other);
        },
        componentWillMount: function componentWillMount() {
            this.spreadConfig();
        },
        componentWillReceiveProps: function componentWillReceiveProps() {
            this.spreadConfig();
        }
    };
};

var configMixin = exports.configMixin = customConfigMixin();
